### PR TITLE
Node peering rework

### PR DESCRIPTION
**Pull request by 0xTylerHolmes**
NodePeer implementation with api rework, mock, test, and multi client implementations.

### CODE REVIEW

**Comment by 0xTylerHolmes** [comment]
see #82

File: api/v1/peers.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	"github.com/pkg/errors"
+)
+
+type Peers struct {
+	Peers []Peer
+}
+
+type peersJSON struct {
+	Data []Peer                 `json:"data"`
+	Meta map[string]interface{} `json:"meta,omitempty"`
```
Comment (mcdee) [inline api/v1/peers.go:16]: I think that we should not support this field.  It isn't properly supported across beacon nodes, and the only information it provides can be calculated very easily.

File: http/nodepeers.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	if len(additionalFields) > 0 {
+		request += "?"
+		for _, field := range additionalFields {
+			request += fmt.Sprintf("%s&", field)
+		}
+	}
+	//remove the trailing '&'
+	request = strings.TrimSuffix(request, "&")
```
Comment (mcdee) [inline http/nodepeers.go]: This is quite involved; would be easier to do a single `request = fmt.Sprintf("%s?%s", request, strings.Join(additionalFields,"&"))`

File: api/peeropts.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+// PeerOpts are the options for client side peer filtering.
+type PeerOpts struct {
+	// State of the connection (disconnected, connecting, connected, disconnecting)
+	State []string
```
Comment (mcdee) [inline api/peeropts.go:19]: This and `Direction` would be better off as enums (as much as we can make them).

File: api/v1/attesterduty_test.go [hunk 4]
```diff
@@ -18,8 +18,8 @@ import (
 	"testing"
 
 	api "github.com/attestantio/go-eth2-client/api/v1"
-	require "github.com/stretchr/testify/require"
 	"github.com/stretchr/testify/assert"
+	require "github.com/stretchr/testify/require"
```
Comment (mcdee) [inline api/v1/attesterduty_test.go:22]: There seem to be a number of items in here that shouldn't be part of the PR; please can you remove them?

File: api/v1/peers.go (hunk 1 above)
Comment (mcdee) [reply api/v1/peers.go:16]: This would also allow use to return `[]*Peer` rather than the `Peers` struct, which would be more in keeping with the rest of the API.

File: http/nodepeers.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+	for _, directionFilter := range opts.Direction {
+		additionalFields += fmt.Sprintf("direction=%s&", directionFilter)
+	}
+
+	if len(additionalFields) > 0 {
+		request += "?" + additionalFields
+		// remove the trailing '&'
+		request = strings.TrimSuffix(request, "&")
+	}
```
Comment (mcdee) [inline http/nodepeers.go:36]: I would make `additionalFields` a `[]string` and add each query parameter to the array individually.  Then you can easily combine them with a `fmt.Sprintf("%s?%s",request,strings.Join(additionalFields,"&"))`

This reduces allocations from repeatedly appending to the string.

**Review by mcdee — CHANGES_REQUESTED** [review]
Thanks for the rework.  A few areas where either the underlying API is inconsistent or the return data is a little involved, apart from that looking good.

File: http/nodepeers.go (hunk 5 above)
Comment (0xTylerHolmes) [reply http/nodepeers.go:36]: Ah interesting I reverted that as it felt clunky to make a list, iterate through the opts and then at the end check if the list was nonempty then append the strings. Ill revert to what I had if you prefer.

File: api/v1/attesterduty_test.go (hunk 4 above)
Comment (0xTylerHolmes) [reply api/v1/attesterduty_test.go:22]: Ah yes sorry I'm not sure how those made it in via the lint.

File: api/v1/peers.go (hunk 1 above)
Comment (0xTylerHolmes) [reply api/v1/peers.go:16]: > This would also allow use to return `[]*Peer` rather than the `Peers` struct, which would be more in keeping with the rest of the API.

I like this better, I had been using round trip parsing of responses for the tests so I added it to conform. As its one way I think its best to just remove it.

File: api/peeropts.go (hunk 3 above)
Comment (0xTylerHolmes) [reply api/peeropts.go:19]: To do this would break the peeropts(api) / peer(v1) boundary. We can slide the peeropts into the v1 but that kind of sidesteps my interpretation of the refactor.

Comment (mcdee) [reply api/peeropts.go:19]: We can leave it for now.

**Review by mcdee — COMMENTED** [review]
Looks good; please could you address the few outstanding items from the linter, then should be good to go.

**Review by mcdee — APPROVED** [review]
